lib/code-caller-python.js

```diff
@@ -135,18 +213,24 @@ class PythonCaller {
       this._startChild();
     }
 
-    const localOptions = _.defaults(options, {
-      cwd: __dirname,
-      paths: [],
-      timeout: config.questionTimeoutMilliseconds,
-    });
-    const callData = {
-      file,
-      fcn,
-      args,
-      cwd: localOptions.cwd,
-      paths: localOptions.paths,
-    };
+    let cwd;
+    const paths = [path.join(__dirname, '..', 'python')];
+    if (type === 'question') {
+      cwd = path.join(this.coursePath, 'questions', directory);
+      paths.push(path.join(this.coursePath, 'serverFilesCourse'));
+    } else if (type === 'course-element') {
+      cwd = path.join(this.coursePath, 'elements', directory);
+      paths.push(path.join(this.coursePath, 'serverFilesCourse'));
+    } else if (type === 'core-element') {
+      cwd = path.join(__dirname, '..', 'elements', directory);
+    } else if (type === 'restart') {
+      // Doesn't need a working directory
+    } else {
+      callback(new Error(`Unknown function call type: ${type}`));
+      return;
+    }
```

Comment:
It feels a bit low-level having all this path calculation done in code-caller-python. Is this being done so that freeform doesn't need to know about paths and can thus use either code-caller-python or code-caller-docker transparently?

Reply:
Ok, sounds reasonable. I don't like code-caller-* knowing about paths, but I'd like it even less if freeform had to know about them.